SMLHelper/Patchers/PingTypePatcher.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Harmony;
+using SMLHelper.V2.Handlers;
+using SMLHelper.V2.Utility;
+
+namespace SMLHelper.V2.Patchers
+{
+    internal class PingTypePatcher
+    {
+        private const string TechTypeEnumName = "PingType";
+        internal static readonly int startingIndex = 10;
+        internal static readonly List<int> bannedIndices = new List<int>();
+        private static readonly Dictionary<PingType, Atlas.Sprite> cachedSprites = new Dictionary<PingType, Atlas.Sprite>();
+
+        internal static readonly EnumCacheManager<PingType> cacheManager =
+            new EnumCacheManager<PingType>(
+                enumTypeName: TechTypeEnumName,
+                startingIndex: startingIndex,
+                bannedIDs: ExtBannedIdManager.GetBannedIdsFor(TechTypeEnumName, bannedIndices, PreRegisteredCraftTreeTypes()));
+
+        private static List<int> PreRegisteredCraftTreeTypes()
+        {
+            var declaredPingTypes = (PingType[]) Enum.GetValues(typeof(PingType));
+            var bannedIndices = declaredPingTypes.Select(t => (int) t)
+                .Where(t => t > startingIndex)
+                .Distinct()
+                .ToList();
+            
+            Logger.Log($"Finished known PingType exclusion. {bannedIndices.Count} IDs were added in ban list.");
+            return bannedIndices;
+        }
+
+        internal static PingType AddPingType(string name, Atlas.Sprite sprite)
+        {
+            var cache = cacheManager.RequestCacheForTypeName(name) ?? new EnumTypeCache()
+            {
+                Name = name,
+                Index = cacheManager.GetNextAvailableIndex()
+            };
+
+            if (cacheManager.IsIndexAvailable(cache.Index))
+                cache.Index = cacheManager.GetNextAvailableIndex();
+
+            var pingType = (PingType) cache.Index;
+            cacheManager.Add(pingType, cache.Index, cache.Name);
+            
+            cachedSprites.Add(pingType, sprite);
```

Comment:
This should probably interact with the [`SpritePatcher`](https://github.com/SubnauticaModding/SMLHelper/blob/master/SMLHelper/Patchers/SpritePatcher.cs) instead of handling the sprites here.